File: Python3/1376_time_needed_to_inform_all_employees.py
# https://leetcode.cn/problems/time-needed-to-inform-all-employees/

class Solution:
    def numOfMinutes(self, n, headID, manager, informTime):
        """
        Inputs:
            n, headID: int
            manager, informTime: list[int]
        Outputs:
            res: int
        """
        graph = defaultdict(list)
        for i in range(n):
            graph[manager[i]].append(i)

        res = 0
        q = [headID]
        minutes = [0 for i in range(n)]
        while q:
            length = len(q)
            for i in range(length):
                cur = q.pop(0)
                for employee in graph[cur]:
                    q.append(employee)
                    minutes[employee] = minutes[cur] + informTime[cur]
                    if res < minutes[employee]:
                        res = minutes[employee]
        return res

File: Cpp/1376_time_needed_to_inform_all_employees.cpp
// https://leetcode.cn/problems/time-needed-to-inform-all-employees/

class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        unordered_map<int, vector<int>> graph;
        for (int i = 0; i < n; ++i) {
            graph[manager[i]].push_back(i);
        }

        int res = 0;
        queue<int> q;
        q.push(headID);
        vector<int> minutes(n, 0);
        while (!q.empty()) {
            int len = q.size();
            for (int i = 0; i < len; ++i) {
                int cur = q.front();
                q.pop();
                for (int employee: graph[cur]) {
                    q.push(employee);
                    minutes[employee] = minutes[cur] + informTime[cur];
                    if (res < minutes[employee]) {
                        res = minutes[employee];
                    }
                }
            }
        }
        return res;
    }
};
